Hammerfest.Server/Dns/DnsService.cs

```diff
@@ -1,61 +1,48 @@
+using System.Collections.Immutable;
 using Hammerfest.Server.Env;
 using Microsoft.Extensions.Options;
 
 namespace Hammerfest.Server.Dns;
 
-public record DnsService(ILogger<DnsService> Logger, IOptions<DnsOptions> Options, ISystemEnvironment Environment) : IHostedService
+public record DnsService
+    (ILogger<DnsService> Logger, IOptions<DnsOptions> Options, ISystemEnvironment Environment) : IHostedService
 {
-    public const string ServServ = "servserv.generals.ea.com";
-    private const string ServerIp = "127.0.0.1";
+    public static ImmutableArray<string> Domains = ImmutableArray.Create(
+        "servserv.generals.ea.com",
+        "cc3xp1.available.gamespy.com"
+    );
+
+    private Dictionary<string, string?> _modifications = new();
+
+    private const string? ServerIp = "127.0.0.1";
 
     private readonly object _stateLock = new();
-    private string? _previousIpAddress;
-    private bool _hostsFileChanged;
 
     public Task StartAsync(CancellationToken cancellationToken)
     {
         if (!Options.Value.Enabled) return Task.CompletedTask;
 
-        var entry = HostsFile.FindIpAddress(Environment, ServServ);
-        if (entry != ServerIp)
+        var newEntries = Domains.ToDictionary(domain => domain, ip => ServerIp);
+        lock (_stateLock)
         {
-            if (entry != null)
-                HostsFile.ReplaceEntry(Environment, ServerIp, ServServ);
-            else
-                HostsFile.AddEntry(Environment, ServerIp, ServServ);
-
-            Logger.LogInformation(
-                """Temporarily wrote entry for host "{Host}" in file "{HostFile}" with address "{IpAddress}".""",
-                ServServ, Environment.HostsFilePath, ServerIp);
-
-            lock (_stateLock)
-            {
-                _previousIpAddress = entry;
-                _hostsFileChanged = true;
-            }
+            _modifications = HostsFile.ModifyEntries(Environment, newEntries);
         }
 
+        Logger.LogInformation("Successfully overwritten original DNS entries with the server entries");
+        Logger.LogDebug(string.Join(System.Environment.NewLine, newEntries));
+
         return Task.CompletedTask;
     }
 
     public Task StopAsync(CancellationToken cancellationToken)
     {
-        string? previousIpAddress;
         lock (_stateLock)
         {
-            if (!_hostsFileChanged) return Task.CompletedTask;
-            previousIpAddress = _previousIpAddress;
+            _modifications = HostsFile.ModifyEntries(Environment, _modifications);
```

Comment:
Here, you've lost a very important aspect of behavior.

Previously, we were additionally validating whether the hosts file content is correct according to our expectations, and we were not touching it if it was externally changed. Consider this situation now:
1. The host file has no entries.
2. `StartAsync` fills it with the entries.
3. The user goes to the file and changes one entry manually.
   From that moment, I consider _that one entry_ user-managed and proclaim that we should not touch it when doing rollback in `StopAsync`.
4. On `StopAsync`, we should only roll back (remove or replace as it was) the single entry that was _not_ changed by the user in the meantime.

Please update the code to reflect this behavior, add a test for the `DnsService` and a test for the `HostsFile` (I think that a new function called `Rollback` would be needed in the `HostsFile` for that, and it should accept a new dictionary parameter with the entries for verification).

You may choose to update the existing `ModifyEntries` method with a new verification parameter; this is up to you.